pkg/node/heartbeat/server.go

```diff
@@ -0,0 +1,96 @@
+package heartbeat
+
+import (
+	"context"
+	"time"
+
+	"github.com/bacalhau-project/bacalhau/pkg/lib/collections"
+	natsPubSub "github.com/bacalhau-project/bacalhau/pkg/nats/pubsub"
+	"github.com/bacalhau-project/bacalhau/pkg/pubsub"
+
+	"github.com/nats-io/nats.go"
+	"github.com/rs/zerolog/log"
+)
+
+type HeartbeatServer struct {
+	subscription *natsPubSub.PubSub[Heartbeat]
+	pqueue       *collections.HashedPriorityQueue[string, Heartbeat]
+}
+
+func NewServer(conn *nats.Conn) (*HeartbeatServer, error) {
+	subParams := natsPubSub.PubSubParams{
+		Subject: heartbeatTopic,
+		Conn:    conn,
+	}
+
+	subscription, err := natsPubSub.NewPubSub[Heartbeat](subParams)
+	if err != nil {
+		return nil, err
+	}
+
+	pqueue := collections.NewHashedPriorityQueue[string, Heartbeat](
+		func(h Heartbeat) string {
+			return h.NodeID
+		},
+	)
+
+	return &HeartbeatServer{subscription: subscription, pqueue: pqueue}, nil
+}
+
+func (h *HeartbeatServer) Start(ctx context.Context) error {
+	if err := h.subscription.Subscribe(ctx, h); err != nil {
+		return err
+	}
+
+	go func(ctx context.Context) {
+		log.Ctx(ctx).Info().Msg("Heartbeat server started")
+		<-ctx.Done()
+		_ = h.subscription.Close(ctx)
+		log.Ctx(ctx).Info().Msg("Heartbeat server shutdown")
+	}(ctx)
```

Comment:
remove the go routine and call `h.subscription.Close(ctx)` in a deferred function.

Reply:
But then it'll close it when the function returns, and I don't want to close it until the context is cancelled.  Deliberately didn't run this Start method in it's own goroutine so that I can return the error if setup fails.

Reply:
I mean a deferred function in the go routine below this one that also handles context cancellation. Wouldn't that work?
e.g.
```go
func (h *HeartbeatServer) Start(ctx context.Context) error {
	if err := h.subscription.Subscribe(ctx, h); err != nil {
		return err
	}
       	log.Ctx(ctx).Info().Msg("Heartbeat server started")
       	
       	go func(ctx context.Context) {
       	        defer func() {
       	                h.subscription.Close(ctx) // TODO and maybe handle error with log message
       	                log.Ctx(ctx).Info().Msg("Heartbeat server shutdown")
       	        }
		ticker := time.NewTicker(heartbeatQueueCheckFrequency)

		for {
			select {
			case <-ctx.Done():
				return
			case <-ticker.C:
				// We'll iterate over the priority queue and check if the last heartbeat
				// received from each node is older than 10 seconds. If it is, we'll
				// consider the node as dead.
			}
		}
	}(ctx)

	return nil
}

Reply:
Oh yeah, got it. I think I refactored the other goroutine to no longer being necessary but forgot to remove it.